File: app/src/main/java/com/example/chris/atp_music_player/adapters/SongListAdapter.java
package com.example.chris.atp_music_player.adapters;

import android.content.Context;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import com.example.chris.atp_music_player.R;
import com.example.chris.atp_music_player.models.Song;
import com.example.chris.atp_music_player.ui.MainActivity;

import java.util.List;

import butterknife.ButterKnife;
import butterknife.InjectView;
import io.realm.Realm;

public class SongListAdapter extends RecyclerView.Adapter<SongListAdapter.ViewHolder> {
    private Context mContext;
    private List<Song> mSongList;

    public SongListAdapter(Context context, List<Song> songList) {
        mContext = context;
        mSongList = songList;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int position) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_song_list, parent, false);
        ViewHolder viewHolder = new ViewHolder(view);

        viewHolder.mOverflow.setOnClickListener(new OnSongOverflowMenuClicked(viewHolder, mSongList, mContext));

        return viewHolder;
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, int position) {
        viewHolder.bind(mSongList.get(position));
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        @InjectView(R.id.item_song_list_title)
        TextView mTitle;
        @InjectView(R.id.item_song_list_subtitle)
        TextView mSubtitle;
        @InjectView(R.id.item_song_overflow)
        ImageButton mOverflow;

        public ViewHolder(View view) {
            super(view);

            view.setOnClickListener(this);
            ButterKnife.inject(this, view);
        }

        public void bind(Song song) {
            mTitle.setText(song.getTitle());
            mSubtitle.setText(song.getArtist());
        }

        @Override
        public void onClick(View view) {
            ((MainActivity) mContext).pushMedia(mSongList, getAdapterPosition());
        }
    }

    @Override
    public int getItemCount() {
        return mSongList.size();
    }
}

File: app/src/main/java/com/example/chris/atp_music_player/adapters/SongListAdapter.kt
package com.example.chris.atp_music_player.adapters

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.chris.atp_music_player.R
import com.example.chris.atp_music_player.Song
import kotlinx.android.synthetic.main.item_song_list.view.*

class SongListAdapter(val songs: List<Song>, val onClick: (List<Song>, Int) -> Unit) :
        RecyclerView.Adapter<SongListAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SongListAdapter.ViewHolder? {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_song_list, parent, false)
        return ViewHolder(view, songs, onClick)
    }

    override fun onBindViewHolder(holder: SongListAdapter.ViewHolder, position: Int) {
        holder.bind(songs[position])
    }

    override fun getItemCount() = songs.size

    class ViewHolder(view: View, val songs: List<Song>, val onClick: (List<Song>, Int) -> Unit) : RecyclerView.ViewHolder(view) {
        fun bind(song: Song) {
            with(song) {
                itemView.title.text = title
                itemView.subtitle.text = artist
                itemView.setOnClickListener { onClick(songs, adapterPosition) }
            }
        }
    }
}